File: scraper_service/hack_scraper/cleanup_expired.py
import pymongo
from hack_scraper.date_utils import is_hackathon_closed

def cleanup_expired_hackathons(mongo_uri="mongodb://localhost:27017", db_name="hackathons", collection_name="hackathons"):
    """
    Connects to MongoDB and deletes all hackathon documents whose deadline date or title indicates it is expired.
    """
    try:
        client = pymongo.MongoClient(mongo_uri, serverSelectionTimeoutMS=3000)
        db = client[db_name]
        collection = db[collection_name]
        
        cursor = collection.find({})
        expired_ids = []
        deleted_titles = []

        for item in cursor:
            deadline = item.get("deadline", "")
            title = item.get("title", "")
            if is_hackathon_closed(deadline, title):
                expired_ids.append(item["_id"])
                deleted_titles.append(title)

        if expired_ids:
            res = collection.delete_many({"_id": {"$in": expired_ids}})
            print(f"[Python Cleanup] Removed {res.deleted_count} expired/closed hackathons from MongoDB.")
            return res.deleted_count
        else:
            print("[Python Cleanup] No expired hackathons found in MongoDB.")
            return 0
    except Exception as e:
        print(f"[Python Cleanup Error] Could not clean expired hackathons: {e}")
        return 0

if __name__ == "__main__":
    cleanup_expired_hackathons()

File: backend/src/utils/cleanupExpired.js
const Hackathon = require('../models/Hackathon');
const { isHackathonClosed } = require('./dateUtils');

/**
 * Sweeps the hackathons collection in MongoDB and removes any hackathon
 * whose registration is closed or deadline has passed compared to current date.
 * 
 * @param {Date} [referenceDate] - Optional reference date (defaults to current date)
 * @returns {Promise<{ totalScanned: number, totalDeleted: number, deletedTitles: string[] }>}
 */
async function cleanupExpiredHackathons(referenceDate = new Date()) {
  try {
    const hackathons = await Hackathon.find({});
    const expiredIds = [];
    const deletedTitles = [];

    for (const item of hackathons) {
      if (isHackathonClosed(item.deadline, item.title, referenceDate)) {
        expiredIds.push(item._id);
        deletedTitles.push(item.title);
      }
    }

    if (expiredIds.length > 0) {
      await Hackathon.deleteMany({ _id: { $in: expiredIds } });
      console.log(`[Auto-Cleanup] Successfully removed ${expiredIds.length} expired/closed hackathons.`);
    } else {
      console.log(`[Auto-Cleanup] Clean sweep complete. No expired hackathons found.`);
    }

    return {
      totalScanned: hackathons.length,
      totalDeleted: expiredIds.length,
      deletedTitles: deletedTitles.slice(0, 10), // Limit array length in return payload
    };
  } catch (error) {
    console.error(`[Auto-Cleanup Error] Failed to cleanup expired hackathons:`, error);
    throw error;
  }
}

module.exports = {
  cleanupExpiredHackathons,
};
